### ISSUE TITLE
extconf.rb is incorrect for Windows

### ISSUE THREAD

**Issue by badal** [ryanmelt/qtbindings#121]
Building on Windows raises many errors. The first one is due to  extconf.rb with contains a lot of
```mkdir qtbin\<subdir>```
without having ```mkdir qtbin``` before.

Did someone really succeeded building qtbindings (4.8.5.2) on Windows (7), ruby 2.1 or 2.2 ?

It would be really nice to have a binary gem.
_md

**Comment by jmthomas** [comment]
There is a binary gem for Windows: gem install qtbindings

If you want to build there are a lot of prerequisites. The information is rather spread out among the README so read everything and post back specific issues once you have everything installed correctly.

**Comment by badal** [comment]
directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\qmltooling
mkdir: cannot create directory `qtbin\\plugins\\qmltooling': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\sqldrivers
mkdir: cannot create directory `qtbin\\plugins\\sqldrivers': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir lib\2.1
cd ext\build && rmdir /S /Q CMakeFiles
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q generator
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q smoke
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q ruby
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && del /F /Q *
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)

make "DESTDIR="
mkdir ext\build
mkdir: cannot create directory `ext\\build': File exists
make: [makedirs] Error 1 (ignored)
mkdir bin\2.1
mkdir: cannot create directory `bin\\2.1': File exists
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins
mkdir: cannot create directory `qtbin\\plugins': No such file or directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\accessible
mkdir: cannot create directory `qtbin\\plugins\\accessible': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\bearer
mkdir: cannot create directory `qtbin\\plugins\\bearer': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\codecs
mkdir: cannot create directory `qtbin\\plugins\\codecs': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\designer
mkdir: cannot create directory `qtbin\\plugins\\designer': No such file or direc
tory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\graphicssystems
mkdir: cannot create directory `qtbin\\plugins\\graphicssystems': No such file o
r directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\iconengines
mkdir: cannot create directory `qtbin\\plugins\\iconengines': No such file or di
rectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\imageformats
mkdir: cannot create directory `qtbin\\plugins\\imageformats': No such file or d
irectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\phonon_backend
mkdir: cannot create directory `qtbin\\plugins\\phonon_backend': No such file or
 directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\qmltooling
mkdir: cannot create directory `qtbin\\plugins\\qmltooling': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\sqldrivers
mkdir: cannot create directory `qtbin\\plugins\\sqldrivers': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir lib\2.1
mkdir: cannot create directory `lib\\2.1': File exists
make: [makedirs] Error 1 (ignored)
cd ext\build && rmdir /S /Q CMakeFiles
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q generator
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q smoke
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q ruby
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && del /F /Q *
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
set CC=mingw32-gcc.exe
set CXX=mingw32-g++.exe
cd ext\build && \
cmake -DCMAKE_MINIMUM_REQUIRED_VERSION=2.6 \
-G "MinGW Makefiles" \
-DCMAKE_MAKE_PROGRAM=mingw32-make.exe \
-Wno-dev \
-DRUBY_EXECUTABLE=C:/Ruby21/bin/ruby \
..
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [build] Error 1 (ignored)
cd ext\build && mingw32-make
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: *** [build] Error 1

make failed, exit code 2

Gem files will remain installed in C:/Ruby21/lib/ruby/gems/2.1.0/gems/qtbindings
-4.8.6.1 for inspection.
Results logged to C:/Ruby21/lib/ruby/gems/2.1.0/extensions/x86-mingw32/2.1.0/qtb
indings-4.8.6.1/gem_make.out

C:\Ruby21\bin>

C:\Ruby21\bin>gem install qtbindings --plaform=windows
ERROR:  While executing gem ... (OptionParser::InvalidOption)
    invalid option: --plaform=windows

C:\Ruby21\bin>gem install qtbindings
Building native extensions.  This could take a while...
ERROR:  Error installing qtbindings:
        ERROR: Failed to build gem native extension.

    C:/Ruby21/bin/ruby.exe extconf.rb

make "DESTDIR=" clean
mkdir ext\build
mkdir bin\2.1
mkdir qtbin\plugins
mkdir: cannot create directory `qtbin\\plugins': No such file or directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\accessible
mkdir: cannot create directory `qtbin\\plugins\\accessible': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\bearer
mkdir: cannot create directory `qtbin\\plugins\\bearer': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\codecs
mkdir: cannot create directory `qtbin\\plugins\\codecs': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\designer
mkdir: cannot create directory `qtbin\\plugins\\designer': No such file or direc
tory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\graphicssystems
mkdir: cannot create directory `qtbin\\plugins\\graphicssystems': No such file o
r directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\iconengines
mkdir: cannot create directory `qtbin\\plugins\\iconengines': No such file or di
rectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\imageformats
mkdir: cannot create directory `qtbin\\plugins\\imageformats': No such file or d
irectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\phonon_backend
mkdir: cannot create directory `qtbin\\plugins\\phonon_backend': No such file or
 directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\qmltooling
mkdir: cannot create directory `qtbin\\plugins\\qmltooling': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\sqldrivers
mkdir: cannot create directory `qtbin\\plugins\\sqldrivers': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir lib\2.1
cd ext\build && rmdir /S /Q CMakeFiles
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q generator
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q smoke
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q ruby
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && del /F /Q *
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)

make "DESTDIR="
mkdir ext\build
mkdir: cannot create directory `ext\\build': File exists
make: [makedirs] Error 1 (ignored)
mkdir bin\2.1
mkdir: cannot create directory `bin\\2.1': File exists
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins
mkdir: cannot create directory `qtbin\\plugins': No such file or directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\accessible
mkdir: cannot create directory `qtbin\\plugins\\accessible': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\bearer
mkdir: cannot create directory `qtbin\\plugins\\bearer': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\codecs
mkdir: cannot create directory `qtbin\\plugins\\codecs': No such file or directo
ry
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\designer
mkdir: cannot create directory `qtbin\\plugins\\designer': No such file or direc
tory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\graphicssystems
mkdir: cannot create directory `qtbin\\plugins\\graphicssystems': No such file o
r directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\iconengines
mkdir: cannot create directory `qtbin\\plugins\\iconengines': No such file or di
rectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\imageformats
mkdir: cannot create directory `qtbin\\plugins\\imageformats': No such file or d
irectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\phonon_backend
mkdir: cannot create directory `qtbin\\plugins\\phonon_backend': No such file or
 directory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\qmltooling
mkdir: cannot create directory `qtbin\\plugins\\qmltooling': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir qtbin\plugins\sqldrivers
mkdir: cannot create directory `qtbin\\plugins\\sqldrivers': No such file or dir
ectory
make: [makedirs] Error 1 (ignored)
mkdir lib\2.1
mkdir: cannot create directory `lib\\2.1': File exists
make: [makedirs] Error 1 (ignored)
cd ext\build && rmdir /S /Q CMakeFiles
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q generator
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q smoke
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && rmdir /S /Q ruby
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
cd ext\build && del /F /Q *
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [clean] Error 1 (ignored)
set CC=mingw32-gcc.exe
set CXX=mingw32-g++.exe
cd ext\build && \
cmake -DCMAKE_MINIMUM_REQUIRED_VERSION=2.6 \
-G "MinGW Makefiles" \
-DCMAKE_MAKE_PROGRAM=mingw32-make.exe \
-Wno-dev \
-DRUBY_EXECUTABLE=C:/Ruby21/bin/ruby \
..
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: [build] Error 1 (ignored)
cd ext\build && mingw32-make
/usr/bin/sh: line 0: cd: extbuild: No such file or directory
make: *** [build] Error 1

make failed, exit code 2

Gem files will remain installed in C:/Ruby21/lib/ruby/gems/2.1.0/gems/qtbindings
-4.8.6.1 for inspection.
Results logged to C:/Ruby21/lib/ruby/gems/2.1.0/extensions/x86-mingw32/2.1.0/qtb
indings-4.8.6.1/gem_make.out

C:\Ruby21\bin>

C:\Ruby21\bin>cd ..\..\Ruby22\bin

C:\Ruby22\bin>gem install qtbindings
Fetching: qtbindings-qt-4.8.6-x86-mingw32.gem (100%)
Successfully installed qtbindings-qt-4.8.6-x86-mingw32
Fetching: qtbindings-4.8.6.1-x86-mingw32.gem (100%)
Successfully installed qtbindings-4.8.6.1-x86-mingw32
Parsing documentation for qtbindings-qt-4.8.6-x86-mingw32
Installing ri documentation for qtbindings-qt-4.8.6-x86-mingw32
Parsing documentation for qtbindings-4.8.6.1-x86-mingw32
Installing ri documentation for qtbindings-4.8.6.1-x86-mingw32
Done installing documentation for qtbindings-qt, qtbindings after 4 seconds
2 gems installed

C:\Ruby22\bin>

**Comment by badal** [comment]
Yes, the binary works for Ruby22.
But previously I tried for Ruby21. Still, Ruby21 wants to build a native extension...
```
C:\Ruby22\bin>gem install qtbindings
Fetching: qtbindings-qt-4.8.6-x86-mingw32.gem (100%)
Successfully installed qtbindings-qt-4.8.6-x86-mingw32
Fetching: qtbindings-4.8.6.1-x86-mingw32.gem (100%)
Successfully installed qtbindings-4.8.6.1-x86-mingw32
Parsing documentation for qtbindings-qt-4.8.6-x86-mingw32
Installing ri documentation for qtbindings-qt-4.8.6-x86-mingw32
Parsing documentation for qtbindings-4.8.6.1-x86-mingw32
Installing ri documentation for qtbindings-4.8.6.1-x86-mingw32
Done installing documentation for qtbindings-qt, qtbindings after 4 seconds
2 gems installed

C:\Ruby22\bin>cd ..\..\Ruby21\bin

C:\Ruby21\bin>gem install qtbindings
Building native extensions.  This could take a while...
ERROR:  Error installing qtbindings:
        ERROR: Failed to build gem native extension.
....

```
>

**Comment by ryanatball** [comment]
I've installed the binary gem with Ruby21 many times.    What version of gem are you running: 
```gem --version```

**Comment by badal** [comment]
2.2.3

**Comment by badal** [comment]
update rubygems : it works now !

Also did chcp 65001

**Comment by badal** [comment]
Thks.
Sorry for the noise.

**Comment by ryanatball** [comment]
No problem.   It looks like creating the 64-bit version of the gem breaks installs with older versions of rubygems.   Good to know.

**Comment by badal** [comment]
OK
BTW, qtbindings is a wonderful gem !
You must have seen that, built on qtbindings:

http://cosmosrb.com/

**Comment by badal** [comment]
Of course, you made it ....

**Comment by ryanatball** [comment]
:)  Thanks for the compliments.  Are you using COSMOS?

**Comment by badal** [comment]
No. Just got the info by some ruby site.

I use qtbindings for the management system (pure ruby, no rails) of the French Mathematical Society. I developed it and maintain it on Windows, but it runs on OSX server. I also have a VirtualBox implementation on CentOS.

I tried FOX, GTK, ... but only Qt gives exactly the same aspect on all environments.

Oh, yes, I am a mathematician, and a frenchman, and retired (I am 78, and maybe the older ruby developer...).
